src/sorta.nim

```diff
@@ -409,45 +414,79 @@ proc `$`*[Key, Val](b: SortedTable[Key, Val]): string =
-      currentNode = currentNode.p0
-
-  var
-    root = b.root
-    currentEntry: Entry[Key, Val]
-    stack = newSeq[Entry[Key, Val]]()
-
-  stack.populateWithLeftmostChildren(root)
-  while stack.len > 0:
-    currentEntry = stack.pop()
-    yield currentEntry
-    stack.populateWithLeftmostChildren(currentEntry.p)
-
-
-iterator mentries[Key, Val](b: var SortedTable[Key, Val]): var Entry[Key, Val] =
-  template populateWithLeftmostChildren(stack, cntNode) =
-    var currentNode = cntNode
-    while not currentNode[].isNil:
-      for i in countdown(currentNode[].m-1, 0):
-        stack.add(addr currentNode[].e[i])
-      currentNode = addr currentNode[].p0
-
-  var
-    root = unsafeAddr b.root
-    currentEntry: ptr Entry[Key, Val]
-    stack = newSeq[ptr Entry[Key, Val]]()
-
-  stack.populateWithLeftmostChildren(root)
-  while stack.len > 0:
-    currentEntry = stack.pop()
-    yield currentEntry[]
-    stack.populateWithLeftmostChildren(addr currentEntry[].p)
-
+proc key[Key, Val](cursor: CursorPosition[Key, Val]): Key =
+  ## Return the key for a given cursor position.
+  cursor.node.e[cursor.entry].key
+
+proc val[Key, Val](cursor: CursorPosition[Key, Val]): Val =
+  ## Return the value for a given cursor position.
+  cursor.node.e[cursor.entry].val
+
+proc mval[Key, Val](cursor: CursorPosition[Key, Val]): var Val =
+  ## Returns a reference to the value for a given cursor position.
+  cursor.node.e[cursor.entry].val
+
+proc search[Key, Val](b: SortedTable[Key, Val], key: Key): Cursor[Key, Val] =
+  ## Calculates the cursor pointing to the given key.
+  var a = b.root
+  while not a.isNil:
+    var r = binarySearch(key, a)
+    if r < a.m:
+      result.stack.add((a, r))
+      if eq(key, a.e[r].key):
+        break
+    a = if r == 0: a.p0 else: a.e[r-1].p
+  # add a dummy entry for first next call
+  result.stack.add((nil, 0))
+
+proc current[Key, Val](cursor: Cursor[Key, Val]): CursorPosition[Key, Val] =
+  ## Returns the current position of a cursor.
+  ## This call is only valid if cursor.next previously returned true.
+  cursor.stack[^1]
+
+proc next[Key, Val](cursor: var Cursor[Key, Val]): bool =
+  ## Moves the cursor forward returning true if cursor.current is now valid.
+  ## Never call current after next returns false.
+  var (node, oldEntry) = cursor.stack.pop()
+  if not node.isNil:
+    var newEntry = oldEntry + 1
+    if newEntry < node.m:
+        cursor.stack.add((node, newEntry))
+    var child = node.e[oldEntry].p
+    if not child.isNil:
+      while not child.isNil:
+        cursor.stack.add((child, 0))
+        child = child.p0
+  return cursor.stack.len > 0
+
+proc cursorFromStart[Key, Val](b: SortedTable[Key, Val]): Cursor[Key, Val] =
+  result = Cursor[Key,Val](stack: @[])
+  var a = b.root
+  while not a.isNil:
+    result.stack.add((a, 0))
+    a = a.p0
+  result.stack.add((nil, 0))
+
+iterator scanFrom*[Key, Val](b: SortedTable[Key, Val], fromKey: Key): tuple[key: Key, val: Val] =
```

Comment:
Since this is similar to `pairs` iterator (it yields key-value tuples), maybe a more meaningful name would be `pairsFrom`?

And then the `scan` iterator could be `pairsBetween`?

Reply:
Done